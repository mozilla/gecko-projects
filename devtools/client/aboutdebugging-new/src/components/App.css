/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/*
 * The current layout of about:debugging is
 *
 *  +-------------+-------------------------------+
 *  |   Sidebar   | Page (Runtime or Connect)     |
 *  |   (240px)   |                               |
 *  |             |                               |
 *  +-------------+-------------------------------+
 */
.app {
  display: grid;
  font-size: 15px;
  grid-column-gap: 40px;
  grid-template-columns: 240px auto;
  height: 100vh;
  overflow: hidden;
  width: 100vw;
}

.page {
  padding-block-start: 60px;
  padding-inline-end: 40px;
  overflow-y: auto;
}
