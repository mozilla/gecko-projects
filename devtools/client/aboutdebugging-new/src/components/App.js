/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { connect } = require("devtools/client/shared/vendor/react-redux");
const { createFactory, PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const Route = createFactory(require("devtools/client/shared/vendor/react-router-dom").Route);
const Switch = createFactory(require("devtools/client/shared/vendor/react-router-dom").Switch);
const Redirect = createFactory(require("devtools/client/shared/vendor/react-router-dom").Redirect);

const Types = require("../types/index");
const { RUNTIMES } = require("../constants");

const ConnectPage = createFactory(require("./connect/ConnectPage"));
const RuntimePage = createFactory(require("./RuntimePage"));
const Sidebar = createFactory(require("./sidebar/Sidebar"));

class App extends PureComponent {
  static get propTypes() {
    return {
      adbAddonStatus: Types.adbAddonStatus,
      // The "dispatch" helper is forwarded to the App component via connect.
      // From that point, components are responsible for forwarding the dispatch
      // property to all components who need to dispatch actions.
      dispatch: PropTypes.func.isRequired,
      // getString prop is injected by the withLocalization wrapper
      getString: PropTypes.func.isRequired,
      isScanningUsb: PropTypes.bool.isRequired,
      networkEnabled: PropTypes.bool.isRequired,
      networkLocations: PropTypes.arrayOf(Types.location).isRequired,
      networkRuntimes: PropTypes.arrayOf(Types.runtime).isRequired,
      selectedPage: Types.page,
      selectedRuntimeId: PropTypes.string,
      usbRuntimes: PropTypes.arrayOf(Types.runtime).isRequired,
      wifiEnabled: PropTypes.bool.isRequired,
    };
  }

  componentDidUpdate() {
    this.updateTitle();
  }

  updateTitle() {
    const { getString, selectedPage, selectedRuntimeId } = this.props;

    const runtimeTitle = selectedRuntimeId ?
                          getString(
                            "about-debugging-page-title-with-runtime",
                            { selectedPage, selectedRuntimeId }
                          )
                          : getString(
                            "about-debugging-page-title",
                            { selectedPage }
                          );

    document.title = runtimeTitle;
  }

  renderConnect() {
    const {
      adbAddonStatus,
      dispatch,
      networkEnabled,
      networkLocations,
      wifiEnabled,
    } = this.props;

    return ConnectPage({
      adbAddonStatus,
      dispatch,
      networkEnabled,
      networkLocations,
      wifiEnabled,
    });
  }

  // The `match` object here is passed automatically by the Route object.
  // We are using it to read the route path.
  // See react-router docs:
  // https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/match.md
  renderRuntime({ match }) {
    // Redirect to This Firefox in these cases:
    // - If the runtimepage for a device is the first page shown (since we can't
    //   keep connections open between page reloads).
    // - If no runtimeId is given.
    // - If runtime is not in the runtimes list or disconnected (this is handled later)
    const isDeviceFirstPage =
      !this.props.selectedPage &&
      match.params.runtimeId !== RUNTIMES.THIS_FIREFOX;
    if (!match.params.runtimeId || isDeviceFirstPage) {
      return Redirect({ to: `/runtime/${RUNTIMES.THIS_FIREFOX}` });
    }

    const isRuntimeAvailable = id => {
      const runtimes = [
        ...this.props.networkRuntimes,
        ...this.props.usbRuntimes,
      ];
      const runtime = runtimes.find(x => x.id === id);
      return runtime && runtime.runtimeDetails;
    };

    const { dispatch } = this.props;

    let runtimeId = match.params.runtimeId || RUNTIMES.THIS_FIREFOX;
    if (match.params.runtimeId !== RUNTIMES.THIS_FIREFOX) {
      const rawId = decodeURIComponent(match.params.runtimeId);
      if (isRuntimeAvailable(rawId)) {
        runtimeId = rawId;
      } else {
        // Also redirect to "This Firefox" if runtime is not found
        return Redirect({ to: `/runtime/${RUNTIMES.THIS_FIREFOX}` });
      }
    }

    // we need to pass a key so the component updates when we want to showcase
    // a different runtime
    return RuntimePage({ dispatch, key: runtimeId, runtimeId });
  }

  renderRoutes() {
    return Switch(
      {},
      Route({
        path: "/connect",
        render: () => this.renderConnect(),
      }),
      Route({
        path: "/runtime/:runtimeId",
        render: routeProps => this.renderRuntime(routeProps),
      }),
      // default route when there's no match which includes "/"
      // TODO: the url does not match "/" means invalid URL,
      // in this case maybe we'd like to do something else than a redirect.
      // See: https://bugzilla.mozilla.org/show_bug.cgi?id=1509897
      Route({
        render: () => Redirect({ to: "/connect"}),
      })
    );
  }

  render() {
    const {
      adbAddonStatus,
      dispatch,
      isScanningUsb,
      networkRuntimes,
      selectedPage,
      selectedRuntimeId,
      usbRuntimes,
    } = this.props;

    return Localized(
      { },
      dom.div(
        { className: "app" },
        Sidebar({
          adbAddonStatus,
          className: "app__sidebar",
          dispatch,
          isScanningUsb,
          networkRuntimes,
          selectedPage,
          selectedRuntimeId,
          usbRuntimes,
        }),
        dom.main({ className: "app__content" }, this.renderRoutes())
      )
    );
  }
}

const mapStateToProps = state => {
  return {
    adbAddonStatus: state.ui.adbAddonStatus,
    isScanningUsb: state.ui.isScanningUsb,
    networkEnabled: state.ui.networkEnabled,
    networkLocations: state.ui.networkLocations,
    networkRuntimes: state.runtimes.networkRuntimes,
    selectedPage: state.ui.selectedPage,
    selectedRuntimeId: state.runtimes.selectedRuntimeId,
    usbRuntimes: state.runtimes.usbRuntimes,
    wifiEnabled: state.ui.wifiEnabled,
  };
};

const mapDispatchToProps = dispatch => ({
  dispatch,
});

module.exports = FluentReact
  .withLocalization(
      connect(mapStateToProps, mapDispatchToProps)(App));
