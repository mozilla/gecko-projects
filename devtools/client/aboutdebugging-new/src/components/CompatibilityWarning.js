/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { createFactory, PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const Message = createFactory(require("./shared/Message"));

const { MESSAGE_LEVEL } = require("../constants");
const { COMPATIBILITY_STATUS } = require("devtools/client/shared/remote-debugging/version-checker");

const TROUBLESHOOTING_URL = "https://developer.mozilla.org/docs/Tools/WebIDE/Troubleshooting";

const Types = require("../types/index");

class CompatibilityWarning extends PureComponent {
  static get propTypes() {
    return {
      compatibilityReport: Types.compatibilityReport.isRequired,
    };
  }

  render() {
    const { localID, localVersion, minVersion, runtimeID, runtimeVersion, status } =
      this.props.compatibilityReport;

    if (status === COMPATIBILITY_STATUS.COMPATIBLE) {
      return null;
    }

    let localizationId, statusClassName;
    switch (status) {
      case COMPATIBILITY_STATUS.TOO_OLD:
        statusClassName = "js-compatibility-warning-too-old";
        localizationId = "about-debugging-runtime-version-too-old";
        break;
      case COMPATIBILITY_STATUS.TOO_RECENT:
        statusClassName = "js-compatibility-warning-too-recent";
        localizationId = "about-debugging-runtime-version-too-recent";
        break;
      case COMPATIBILITY_STATUS.TOO_OLD_67_DEBUGGER:
        statusClassName = "js-compatibility-warning-too-old-67-debugger";
        localizationId = "about-debugging-runtime-version-too-old-67-debugger";
        break;
    }

    return Message(
      {
        level: MESSAGE_LEVEL.WARNING,
      },
      Localized(
        {
          id: localizationId,
          a: dom.a({
            href: TROUBLESHOOTING_URL,
            target: "_blank",
          }),
          $localID: localID,
          $localVersion: localVersion,
          $minVersion: minVersion,
          $runtimeID: runtimeID,
          $runtimeVersion: runtimeVersion,
        },
        dom.p(
          {
            className: `js-compatibility-warning ${statusClassName}`,
          },
          localizationId,
        ),
      )
    );
  }
}

module.exports = CompatibilityWarning;
