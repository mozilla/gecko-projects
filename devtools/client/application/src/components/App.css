/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/*
 * General styles
 */

a.disabled-link,
a.disabled-link:hover,
a.disabled-link:visited {
  opacity: 0.5;
  cursor: not-allowed;
}

/*
 * The current layout of the application panel is
 *
 *  +---------------------------------------------+
 *  | (header) "Service workers"                  |
 *  +---------------------------------------------+
 *  | Service worker 1                            |
 *  |   (...)                                     |
 *  | Service worker N           (see Worker.css) |
 *  +---------------------------------------------+
 *  |                     Link to about:debugging |
 *  +---------------------------------------------+
 */
.application {
  height: 100vh;
  padding: 0 2rem;
  display: grid;
  -moz-user-select: none;
}

.application--empty {
  align-items: center;
  justify-content: center;
}

.application:not(.application--empty) {
  grid-template-rows: 1fr auto;
}
