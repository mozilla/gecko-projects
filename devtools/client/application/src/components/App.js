/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const { main } = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const LocalizationProvider = createFactory(FluentReact.LocalizationProvider);

const PageSwitcher = createFactory(require("./routing/PageSwitcher"));
const Sidebar = createFactory(require("./routing/Sidebar"));

/**
 * This is the main component for the application panel.
 */
class App extends PureComponent {
  static get propTypes() {
    return {
      fluentBundles: PropTypes.array.isRequired,
    };
  }

  render() {
    const { fluentBundles } = this.props;

    return LocalizationProvider(
      { bundles: fluentBundles },
      main({ className: `app` }, Sidebar({}), PageSwitcher({}))
    );
  }
}

module.exports = App;
