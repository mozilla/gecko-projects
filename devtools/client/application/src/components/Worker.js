/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { Component } = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { a, button, dd, dl, dt, header, li, section, span } = require("devtools/client/shared/vendor/react-dom-factories");
const Services = require("Services");
const { getUnicodeUrl, getUnicodeUrlPath } = require("devtools/client/shared/unicode-url");

loader.lazyRequireGetter(this, "DebuggerClient",
  "devtools/shared/client/debugger-client", true);
loader.lazyRequireGetter(this, "gDevToolsBrowser",
  "devtools/client/framework/devtools-browser", true);

const Strings = Services.strings.createBundle(
  "chrome://devtools/locale/aboutdebugging.properties");

/**
 * This component is dedicated to display a worker, more accurately a service worker, in
 * the list of workers displayed in the application panel. It displays information about
 * the worker as well as action links and buttons to interact with the worker (e.g. debug,
 * unregister, update etc...).
 */
class Worker extends Component {
  static get propTypes() {
    return {
      client: PropTypes.instanceOf(DebuggerClient).isRequired,
      debugDisabled: PropTypes.bool,
      worker: PropTypes.shape({
        active: PropTypes.bool,
        name: PropTypes.string.isRequired,
        scope: PropTypes.string.isRequired,
        // registrationActor can be missing in e10s.
        registrationActor: PropTypes.string,
        workerActor: PropTypes.string
      }).isRequired
    };
  }

  constructor(props) {
    super(props);

    this.debug = this.debug.bind(this);
    this.start = this.start.bind(this);
    this.unregister = this.unregister.bind(this);
  }

  debug() {
    if (!this.isRunning()) {
      console.log("Service workers cannot be debugged if they are not running");
      return;
    }

    let { client, worker } = this.props;
    gDevToolsBrowser.openWorkerToolbox(client, worker.workerActor);
  }

  start() {
    if (!this.isActive() || this.isRunning()) {
      console.log("Running or inactive service workers cannot be started");
      return;
    }

    let { client, worker } = this.props;
    client.request({
      to: worker.registrationActor,
      type: "start"
    });
  }

  unregister() {
    let { client, worker } = this.props;
    client.request({
      to: worker.registrationActor,
      type: "unregister"
    });
  }

  isRunning() {
    // We know the worker is running if it has a worker actor.
    return !!this.props.worker.workerActor;
  }

  isActive() {
    return this.props.worker.active;
  }

  getServiceWorkerStatus() {
    if (this.isActive() && this.isRunning()) {
      return "running";
    } else if (this.isActive()) {
      return "stopped";
    }
    // We cannot get service worker registrations unless the registration is in
    // ACTIVE state. Unable to know the actual state ("installing", "waiting"), we
    // display a custom state "registering" for now. See Bug 1153292.
    return "registering";
  }

  formatScope(scope) {
    let [, remainder] = getUnicodeUrl(scope).split("://");
    return remainder || scope;
  }

  formatSource(source) {
    let parts = source.split("/");
    return getUnicodeUrlPath(parts[parts.length - 1]);
  }

  render() {
    let { worker } = this.props;
    let status = this.getServiceWorkerStatus();

    const unregisterButton = this.isActive() ?
      button({
        onClick: this.unregister,
        className: "devtools-button worker__unregister-button js-unregister-button",
        "data-standalone": true
      },
        Strings.GetStringFromName("unregister"))
      : null;

    const debugLinkDisabled = this.isRunning() ? "" : "worker__debug-link--disabled";
    const debugLink = a({
      onClick: this.isRunning() ? this.debug : null,
      title: this.isRunning() ? null : "Only running service workers can be debugged",
      className: `${debugLinkDisabled} worker__debug-link`
    },
      Strings.GetStringFromName("debug"));

    const startLink = !this.isRunning() ?
      a({ onClick: this.start, className: "worker__start-link" },
        Strings.GetStringFromName("start"))
      : null;

    return li({ className: "worker js-sw-container" },
      header(
        { className: "worker__header" },
        span({ title: worker.scope, className: "worker__scope js-sw-scope" },
          this.formatScope(worker.scope)),
        section(
          { className: "worker__controls" },
          unregisterButton),
      ),
      dl(
        { className: "worker__data" },
        dt({ className: "worker__meta-name" }, "Source"),
        dd({},
            span({ title: worker.scope, className: "js-source-url" },
              this.formatSource(worker.url)),
            debugLink),
        dt({ className: "worker__meta-name" }, "Status"),
        dd({},
          Strings.GetStringFromName(status).toLowerCase(),
          startLink)
      )
    );
  }
}

module.exports = Worker;
