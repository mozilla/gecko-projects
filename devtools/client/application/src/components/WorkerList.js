/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { openTrustedLink } = require("devtools/client/shared/link");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { createFactory, Component } = require("devtools/client/shared/vendor/react");
const { a, article, footer, h1, ul } = require("devtools/client/shared/vendor/react-dom-factories");
const Worker = createFactory(require("./Worker"));

/**
 * This component handles the list of service workers displayed in the application panel
 * and also displays a suggestion to use about debugging for debugging other service
 * workers.
 */
class WorkerList extends Component {
  static get propTypes() {
    return {
      client: PropTypes.object.isRequired,
      workers: PropTypes.object.isRequired,
    };
  }

  render() {
    const { workers, client } = this.props;

    return [
      article({ className: "workers-container" },
        h1({}, "Service Workers"),
        ul({},
          workers.map(worker => Worker({
            client,
            debugDisabled: false,
            worker,
          })))
      ),
      footer({ className: "aboutdebugging-plug" },
        "See about:debugging for Service Workers from other domains",
        a({ className: "aboutdebugging-plug__link",
            onClick: () => openTrustedLink("about:debugging#workers") },
          "Open about:debugging"
        )
      )
    ];
  }
}

// Exports

module.exports = WorkerList;
