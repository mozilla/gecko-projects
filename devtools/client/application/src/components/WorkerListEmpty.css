/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.worker-list-empty {
  max-width: calc(var(--base-unit) * 179);
  font-size: var(--body-10-font-size);
  color: var(--theme-toolbar-color);
}

.worker-list-empty__title {
  font-size: var(--title-20-font-size);
  font-weight: var(--title-20-font-weight);
}

.worker-list-empty__title .external-link {
  font-weight: var(--title-30-font-weight);
} 

.worker-list-empty__tips {
  margin-inline-start: 2rem;
}

.worker-list-empty__tips__item {
  margin-block-start: 0.5rem;
  margin-block-end: 0.5rem;
}
