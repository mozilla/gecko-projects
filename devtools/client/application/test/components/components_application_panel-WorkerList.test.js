/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

// Import libs
const { shallow } = require("enzyme");
const React = require("react");

// Import constants
const {
  CLIENT,
  SINGLE_WORKER_LIST,
  MULTIPLE_WORKER_LIST,
} = require("devtools/client/application/test/components/fixtures/data/constants");

const WorkerList = React.createFactory(
  require("devtools/client/application/src/components/WorkerList")
);

/**
 * Test for workerList.js component
 */
describe("Service Worker List:", () => {
  it("renders the expected snapshot for a workerList with single worker list", () => {
    const wrapper = shallow(
      WorkerList({
        client: CLIENT,
        workers: SINGLE_WORKER_LIST,
        isDebugEnabled: true,
      })
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the expected snapshot for a workerList with multiple workers list", () => {
    const wrapper = shallow(
      WorkerList({
        client: CLIENT,
        workers: MULTIPLE_WORKER_LIST,
        isDebugEnabled: true,
      })
    );
    expect(wrapper).toMatchSnapshot();
  });
});
