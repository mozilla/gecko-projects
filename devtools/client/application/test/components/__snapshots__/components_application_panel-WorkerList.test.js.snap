// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Worker List: renders the expected snapshot for a workerList with multiple workers list 1`] = `
Array [
  <article
    className="workers-container"
    key="workers-container"
  >
    <Localized
      id="serviceworker-list-header"
    >
      <h1
        className="application--title"
      />
    </Localized>
    <ul>
      <Worker
        client={Object {}}
        worker={
          Object {
            "active": true,
            "name": "worker1",
            "registrationFront": "",
            "scope": "SCOPE 123",
            "workerTargetFront": "",
          }
        }
      />
      <Worker
        client={Object {}}
        worker={
          Object {
            "active": false,
            "name": "worker2",
            "registrationFront": "",
            "scope": "SCOPE 456",
            "workerTargetFront": "",
          }
        }
      />
      <Worker
        client={Object {}}
        worker={
          Object {
            "active": true,
            "name": "worker3",
            "registrationFront": "",
            "scope": "SCOPE 789",
            "workerTargetFront": "",
          }
        }
      />
    </ul>
  </article>,
  <Localized
    a={
      <a
        className="aboutdebugging-plug__link"
        onClick={[Function]}
      />
    }
    id="serviceworker-list-aboutdebugging"
    key="serviceworkerlist-footer"
  >
    <footer
      className="aboutdebugging-plug"
    />
  </Localized>,
]
`;

exports[`Service Worker List: renders the expected snapshot for a workerList with single worker list 1`] = `
Array [
  <article
    className="workers-container"
    key="workers-container"
  >
    <Localized
      id="serviceworker-list-header"
    >
      <h1
        className="application--title"
      />
    </Localized>
    <ul>
      <Worker
        client={Object {}}
        worker={
          Object {
            "active": true,
            "name": "worker1",
            "registrationFront": "",
            "scope": "SCOPE 123",
            "workerTargetFront": "",
          }
        }
      />
    </ul>
  </article>,
  <Localized
    a={
      <a
        className="aboutdebugging-plug__link"
        onClick={[Function]}
      />
    }
    id="serviceworker-list-aboutdebugging"
    key="serviceworkerlist-footer"
  >
    <footer
      className="aboutdebugging-plug"
    />
  </Localized>,
]
`;
